Sources/GRPC/ConnectionPool/ConnectionPool.swift

```diff
@@ -564,27 +586,73 @@ extension ConnectionPool: ConnectionManagerConnectivityDelegate {
     default:
       ()
     }
+
+    guard let delegate = self.delegate else { return }
+
+    switch (oldState, newState) {
+    case (.idle, .connecting),
+         (.transientFailure, .connecting):
+      delegate.startedConnecting(id: .init(manager.id))
+
+    case (.connecting, .ready):
+      // The connection becoming ready is handled by 'receivedSettingsMaxConcurrentStreams'.
+      ()
+
+    case (.ready, .idle):
+      delegate.connectionClosed(id: .init(manager.id), error: nil)
+
+    case let (.ready, .transientFailure(error)):
+      delegate.connectionClosed(id: .init(manager.id), error: error)
+
+    case let (.connecting, .transientFailure(error)):
+      delegate.connectFailed(id: .init(manager.id), error: error)
+
+    default:
+      ()
+    }
   }
 
   func connectionIsQuiescing(_ manager: ConnectionManager) {
     self.eventLoop.assertInEventLoop()
-    guard let removed = self._connections.removeValue(forKey: manager.id) else {
-      return
+
+    // Find the relevant connection, mark it as quiescing and drop the connectivity delegate as
+    // these events are no longer relevant.
+    guard let index = self._connections.index(forKey: manager.id) else { return }
+    self._connections.values[index].isQuiescing = true
+    self._connections.values[index].manager.sync.connectivityDelegate = nil
+
+    self.delegate?.connectionQuiescing(id: .init(manager.id))
+
+    // If there's a close future then the underlying channel is open. It will close eventually when
+    // open streams are closed, so drop the H2 delegate and update the pool delegate when that
+    // happens.
+    if let closeFuture = self._connections.values[index].manager.sync.channel?.closeFuture {
```

Comment:
Holy Law of Demeter batman! We absolutely need a clearer spelling for this idea.

Reply:
Agreed. It's certainly more fiddly than I'd like and as it turns out also a little over protective: we will be notified about quiescing only if there are open streams (if there aren't then the connection will just be idled). So the `else` branch here is redundant. I think API for registering an on-close callback on the current connection would probably be sufficient here.